mobile/test/common/integration/client_integration_test.cc

```diff
@@ -325,20 +325,62 @@ TEST_P(ClientIntegrationTest, DirectResponse) {
   // Override to not validate stream intel.
   stream_prototype_->setOnComplete(
       [this](envoy_stream_intel, envoy_final_stream_intel final_intel) {
-        cc_.on_complete_received_byte_count = final_intel.received_byte_count;
-        cc_.on_complete_calls++;
-        cc_.terminal_callback->setReady();
+        cc_->on_complete_received_byte_count = final_intel.received_byte_count;
+        cc_->on_complete_calls++;
+        cc_->terminal_callback->setReady();
       });
 
   default_request_headers_.setHost("127.0.0.1");
   default_request_headers_.setPath("/");
 
   stream_->sendHeaders(envoyToMobileHeaders(default_request_headers_), true);
-  terminal_callback_.waitReady();
-  ASSERT_EQ(cc_.status, "404");
-  ASSERT_EQ(cc_.on_headers_calls, 1);
+  terminal_callback_->waitReady();
+  ASSERT_EQ(cc_->status, "404");
+  ASSERT_EQ(cc_->on_headers_calls, 1);
   stream_.reset();
 }
 
+// like the "Basic" test but with 20 engines
+TEST_P(ClientIntegrationTest, MultipleEngines) {
+  num_engines_for_test_ = 20;
+  initialize();
+
+  Buffer::OwnedImpl request_data = Buffer::OwnedImpl("request body");
+  default_request_headers_.addCopy(AutonomousStream::EXPECT_REQUEST_SIZE_BYTES,
+                                   std::to_string(request_data.length()));
+  for (int i = 0; i < num_engines_for_test_; i++) {
+    multi_stream_prototypes_[i]->setOnData([=](envoy_data c_data, bool end_stream) {
+      if (end_stream) {
+        EXPECT_EQ(Data::Utility::copyToString(c_data), "");
+      } else {
+        EXPECT_EQ(c_data.length, 10);
+      }
+      multi_cc_[i]->on_data_calls++;
+      release_envoy_data(c_data);
+    });
+
+    multi_streams_[i]->sendHeaders(envoyToMobileHeaders(default_request_headers_), false);
+
+    envoy_data c_data = Data::Utility::toBridgeData(request_data);
+    multi_streams_[i]->sendData(c_data);
+
+    Platform::RequestTrailersBuilder builder;
+    std::shared_ptr<Platform::RequestTrailers> trailers =
+        std::make_shared<Platform::RequestTrailers>(builder.build());
+    multi_streams_[i]->close(trailers);
+
+    multi_terminal_callbacks_[i]->waitReady();
+
+    ASSERT_EQ(multi_cc_[i]->on_headers_calls, 1);
+    ASSERT_EQ(multi_cc_[i]->status, "200");
+    ASSERT_EQ(multi_cc_[i]->on_data_calls, 2);
+    ASSERT_EQ(multi_cc_[i]->on_complete_calls, 1);
+    ASSERT_EQ(multi_cc_[i]->on_header_consumed_bytes_from_response, 27);
+    ASSERT_EQ(multi_cc_[i]->on_complete_received_byte_count, 67);
```

Comment:
awesome, looks like you got the multi-engine data request validation working too!

Reply:
Yup! Finally something just worked first try, you love to see it